src/app/data-model/FabricScopedPreEncodedValue.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "FabricScopedPreEncodedValue.h"
+#include <lib/support/CodeUtils.h>
+
+#include <optional>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+FabricScopedPreEncodedValue::FabricScopedPreEncodedValue(const ByteSpan & aData)
+{
+    mReader.Init(aData);
+}
+
+CHIP_ERROR FabricScopedPreEncodedValue::EncodeForRead(TLV::TLVWriter & aWriter, TLV::Tag aTag, FabricIndex aFabricIndex) const
+{
+    // Clone our reader, since we need to do Next() to move to its first element
+    // but not change our state.  And CopyElement needs a non-const reader
+    // anyway since it also changes reader state.
+    TLV::TLVReader reader(mReader);
+
+    ReturnErrorOnFailure(reader.Next());
+
+    return aWriter.CopyElement(aTag, reader);
+}
+
+FabricIndex FabricScopedPreEncodedValue::GetFabricIndex() const
+{
+    TLV::TLVReader reader(mReader);
```

Comment:
Instead of a "copy-like" TLVReader, we could just reinit the reader with the span we have.

Reply:
That's fair.  That's probably simpler, will do that.

Reply:
Done.